kube-runtime/src/scheduler.rs

```diff
@@ -60,12 +68,24 @@ impl<'a, T: Hash + Eq + Clone, R> SchedulerProj<'a, T, R> {
             // Message is already pending, so we can't even expedite it
             return;
         }
-        match self.scheduled.entry(request.message) {
+        match self.scheduled.entry(request.message.clone()) {
             Entry::Occupied(mut old_entry) if old_entry.get().run_at >= request.run_at => {
                 // Old entry will run after the new request, so replace it..
                 let entry = old_entry.get_mut();
+
+                // Compare the new request's reason with the reason we have for the existing
+                // message. If its different then we need to replace the entire message with the new
+                // request's message since we can't update the existing message's reason and we'd
+                // like the emitted message to have the correct reason about emission.
+                if entry.reason != request.reason {
+                    // Something like `queue.update(key, value)` would have been ideal, but alas.
+                    self.queue.try_remove(&entry.queue_key);
+                    entry.queue_key = self.queue.insert_at(request.message, request.run_at);
+                    entry.reason = request.reason;
+                } else {
+                    self.queue.reset_at(&entry.queue_key, request.run_at);
+                }
```

Comment:
isn't it sufficient to just set `entry.reason = request.to_string()` (or whatever generic shortcut we use for it)?
particularly if we only put the reason in the `ScheduleEntry` struct, because then it does not affect the `queue_key` and we should just be able to update the value in place.

the queue popping and re-insertion is definitely more problematic, and we should find another solution for that.

personally, i thought this would just work with the mutable reference we have to the current scheduled `entry`

Reply:
it isnt sufficient to just do `entry.reason = request.to_string()` (that's what i thought at first as well). the `request.message`, i.e. `ReconcileRequest` is stored in the queue, but since the queue is implemented in a generic manner, we can't modify `ReconcileRequest.reason`, which is essentially what we want to do. we want to make sure the `ReconileRequest` object the scheduler emits always has the last observed reason.
an idea i have is to create a trait which has a method like `set_reason()` and then tie `T` to that trait. then in `poll_pop_queue_message` we could simply do:

```rust
let entry = self.scheduled.remove(&msg).expect("");
msg.set_reason(entry.reason)
if can_take_message(&msg) {
    Poll::Ready(msg)
}
```

Reply:
I see :(

I think any solution using the `.remove()` or `.try_remove()` on the `DelayQueue` for then to re-insert it, is a pretty [scary](https://docs.rs/tokio-util/latest/tokio_util/time/delay_queue/struct.DelayQueue.html#panics-2) and confusing solution (although there is already plenty of confusing in there just due to how plain complicated this is).

Was personally hoping to keep the `DelayQueue` for **only** the timing aspect. Basically;

- have `ScheduleRequest<T>` turn into `ScheduleRequest<T, R>` where `R` is some Reason, and `T: ObjectRef<K>`
- this leads to `ObjectRef<K>` in the queue (`DelayQueue<ObjectRef<K>>`) basically bypassing `ReconcileRequest` in scheduler.rs
- then you can only update the `scheduled` hashmap and keep the updateable information there like we do for the `run_at` value (i.e. we move the `reason: R` in there)

...then after looking into this better, I looked again at your [earlier suggested alternative change](https://github.com/kube-rs/kube/commit/844363e740dc98160a7183a721a0dc22b36d99e2#diff-88f29884eed979a9680949f70db6d6c46eb53c8d4346bc1e618fc709697bf34e), and it seems that's basically what you've found also.

That solution does leave `ReconcileRequest` in a somewhat awkward place, given it's public, but you've managed to keep it in a non-breaking way (the alternative would be to break the lower level `applier` + trigger interfaces which i am not keen on). I'm thinking your alternative is probably the way to go forward.